core/shared/src/main/scala/zio/ZIO.scala

```diff
@@ -2022,29 +2022,84 @@ private[zio] trait ZIOFunctions extends Serializable {
    *
    * For a sequential version of this method, see `foreach`.
    */
-  final def foreachPar[R, E, A, B](as: Iterable[A])(fn: A => ZIO[R, E, B]): ZIO[R, E, List[B]] =
-    as.foldRight[ZIO[R, E, List[B]]](effectTotal(Nil)) { (a, io) =>
-        fn(a).zipWithPar(io)((b, bs) => b :: bs)
-      }
-      .refailWithTrace
+  final def foreachPar[R, E, A, B](as: Iterable[A])(fn: A => ZIO[R, E, B]): ZIO[R, E, List[B]] = {
+    def arbiter(
+      promise: Promise[Nothing, Either[Cause[E], List[B]]],
+      buffer: Ref[Array[B]],
+      todo: Ref[Int],
+      idx: Int
+    )(res: Exit[E, B]): ZIO[R, Nothing, Unit] =
+      res.foldM[R, Nothing, Unit](
+        e => promise.succeed(Left(e)).unit,
+        a =>
+          todo.modify { t =>
+            val updateBuffer = buffer.modify(b => (b.update(idx, a), b))
+            if (t == 1) (updateBuffer <* buffer.get.flatMap(b => promise.succeed(Right(b.toList)))) -> 0
+            else updateBuffer                                                                       -> (t - 1)
+          }.flatten
+      )
+
+    (for {
+      size    <- UIO.effectTotal(as.size)
+      todo    <- Ref.make(size)
+      buffer  <- Ref.make(new Array[Any](size).asInstanceOf[Array[B]])
```

Comment:
Should I use `Array.ofDim[B]` and add `ClassTag` to signature or use `ArrayBuffer` or something else?

Reply:
I'll go with `mutable.ArraySeq` because of:

> You would typically use an ArraySeq if you want an array for its performance characteristics, but you also want to create generic instances of the sequence where you do not know the type of the elements and you do not have a ClassTag to provide it at run-time. These issues are explained in the section on arrays.

>